src/main/java/com/aws/greengrass/shadowmanager/ipc/UpdateThingShadowRequestHandler.java

```diff
@@ -178,17 +166,36 @@ public UpdateThingShadowHandlerResponse handleRequest(UpdateThingShadowRequest r
 
                 try {
                     // Generate the new merged document based on the update shadow patch payload.
-                    ShadowDocument updatedDocument = new ShadowDocument(currentDocument);
-                    final Pair<JsonNode, JsonNode> patchStateMetadataPair = updatedDocument
-                            .update(updateDocumentRequest);
+                    final ShadowDocument updatedDocument = new ShadowDocument(currentDocument);
+                    final JsonNode metadata = updatedDocument.update(updateDocumentRequest);
+
+                    // shadow size is based on desired + reported length. Ideally we would just store these separately
+                    // to avoid double serialization, but DB stores the single document
+                    int desiredLength = 0;
+                    int reportedLength = 0;
+                    if (!isNullOrMissing(updatedDocument.getState().getDesired())) {
+                        desiredLength = JsonUtil.getPayloadBytes(updatedDocument.getState().getDesired()).length;
+                    }
+                    if (!isNullOrMissing(updatedDocument.getState().getReported())) {
+                        reportedLength = JsonUtil.getPayloadBytes(updatedDocument.getState().getReported()).length;
+                    }
+
+                    // Make sure new document is not too big
+                    logger.atDebug()
+                            .setEventType(LogEvents.UPDATE_THING_SHADOW.code())
+                            .kv(LOG_THING_NAME_KEY, thingName)
+                            .kv(LOG_SHADOW_NAME_KEY, shadowName)
+                            .kv("mergedShadowStateSize", desiredLength + reportedLength)
```

Comment:
log what our max allowed size is

Reply:
added a log when it loads the config

Reply:
and added an extra warning log when it encounters the problem so it can log that detail